src/search.jl

```diff
@@ -1,81 +1,119 @@
+    cfg::Vector,
+    state::Vector,
+    basis::Vector,
     )
-    perm = partialsortperm(sol.probabilities, 1:k, rev=true)
-    Solution(sol.energies[perm], sol.states[perm], sol.probabilities[perm])
+    tmp = Vector{Int}[]
+    for σ ∈ basis push!(tmp, vcat(state, σ)) end
+    vcat(cfg, tmp)
 end
 
-function _update_solution!(
-    sol::Solution,
-    model::Model, 
-    )
-    prob = eng = psol = [] 
+@inline _init_solution() = (Float64[], Float64[], Vector{Int}[])
 
-    for (i, σ) ∈ enumerate(sol.states)
-        p = conditional_probability(model, σ)
-        push!(prob, sol.probabilities[i] .* p)
-        push!(eng, sol.energies[i] .+ δE(model, σ))
-        # add sol
-    end
+# TODO: logic here can probably be done better
+function _bound(pdo::Vector{Float64}, cut::Int)
+    k = length(pdo)
+    second_phase = false
+    if k > cut + 1 k = cut + 1; second_phase = true end
 
-    sol = Solution(vec(eng), sol, vec(prob)) 
-    _sort(_merge(model, sol), k)
+    idx = partialsortperm(pdo, 1:k, rev=true)
+
+    if second_phase
+        return idx[1:end-1], pdo[last(idx)]
+    else
+        return idx, -Inf
+    end
 end
 
-@inline function _largest_discarded_probability!(
-    lpCut::Float64,
+function _branch_and_bound(
     sol::Solution,
+    network::AbstractGibbsNetwork,
+    node::Int,
+    cut::Int,
     )
-    p = last(sol.probabilities)
-    lpCut < p ? lpCut = p : ()
+
+    # branch
+    pdo, eng, cfg = _init_solution()
+    k = get_prop(network.fg, node, :loc_dim)
+
+    # for (p, σ, e) ∈ zip(sol.probabilities, sol.states, sol.energies)
+    #     pdo = [pdo; p .* conditional_probability(network, σ)]
+    #     eng = [eng; e .+ update_energy(network, σ)]
+    #     cfg = _branch_state(cfg, σ, collect(1:k))
+    #  end
+
+    for (p, σ, e) ∈ zip(sol.probabilities, sol.states, sol.energies)
+        pdo = [pdo; p .* conditional_probability(network, σ)]
+        eng = [eng; e .+ update_energy(network, σ)]
+        cfg = _branch_state(cfg, σ, collect(1:k))
+     end
+
+    # bound
+    K, lp = _bound(pdo, cut)
+    lpCut = sol.largest_discarded_probability
+    lpCut < lp ? lpCut = lp : ()
+
+    Solution(eng[K], cfg[K], pdo[K], lpCut)
 end
 
+#TODO: incorporate "going back" move to improve alghoritm
 function low_energy_spectrum(
-    model::Model, 
-    k::Int
-    )
-    sol = Solution(zeros(k), fill([], k), zeros(k))
-    lpCut = -typemax(Int)
+    network::AbstractGibbsNetwork,
+    cut::Int
+)
+    sol = Solution([0.], [[]], [1.], -Inf)
+
+    perm = zeros(Int, nv(network.fg)) # TODO: to be removed
 
-    for v ∈ 1:model.size 
-        _update_solution!(sol, model)
-        _largest_discarded_probability!(lpCut, sol)
+    #TODO: this should be replaced with the iteration over fg that is consistent with the order network
+    for i ∈ 1:network.i_max, j ∈ 1:network.j_max
+        v_fg = network.map[i, j]
+        perm[v_fg] = j + network.j_max * (i - 1)
+        sol = _branch_and_bound(sol, network, v_fg, cut)
```

Comment:
Given that solution is not mutated by `_branch_and_bound`, can we make it an immutable struct?

Reply:
Done.